File: src/main/services/prompts/dynamic/AvailableTools.ts
import type { PromptModule, PromptContext } from '../PromptTypes'
import { getMcpInstructionRegistry } from '../../mcp/McpInstructionRegistry'

export const AvailableToolsModule: PromptModule = {
  id: 'available-tools',
  layer: 'dynamic',
  priority: 2,
  build: (context: PromptContext) => {
    const lines: string[] = []
    if (context.deferredTools?.length) {
      lines.push('<deferred_tools>')
      lines.push('Use ToolSearch to activate one of these capabilities for the next turn:')
      for (const tool of context.deferredTools) {
        lines.push(`- ${tool.name}: ${tool.summary}`)
      }
      lines.push('</deferred_tools>')
    }
    const mcpInstructions = getMcpInstructionRegistry().render()
    if (mcpInstructions) lines.push(mcpInstructions)
    return lines.join('\n')
  },
}

File: crates/codez-runtime/src/chat/prompt/modules/available_tools.rs
use crate::chat::prompt::types::{PromptContext, PromptLayer, PromptModule};

pub struct AvailableToolsModule;

use crate::chat::prompt::types::BoxFuture;

impl PromptModule for AvailableToolsModule {
    fn id(&self) -> &'static str {
        "available-tools"
    }

    fn layer(&self) -> PromptLayer {
        PromptLayer::Dynamic
    }

    fn priority(&self) -> i32 {
        2
    }

    fn build<'a>(&'a self, ctx: &'a PromptContext) -> BoxFuture<'a, Option<String>> {
        Box::pin(async move {
            let mut lines = Vec::new();

            if let Some(available) = &ctx.available_tools {
                if !available.is_empty() {
                    lines.push("<available_tools>".to_string());
                    lines.push("Tools available in this Provider round:".to_string());
                    for tool in available {
                        lines.push(format!("- {}: {}", tool.name, tool.summary));
                    }
                    lines.push("</available_tools>".to_string());
                }
            }

            if let Some(deferred) = &ctx.deferred_tools {
                if !deferred.is_empty() {
                    lines.push("<deferred_tools>".to_string());
                    lines.push(
                        "Use ToolSearch to activate one of these capabilities for the next turn:"
                            .to_string(),
                    );
                    for tool in deferred {
                        lines.push(format!("- {}: {}", tool.name, tool.summary));
                    }
                    lines.push("</deferred_tools>".to_string());
                }
            }
            if lines.is_empty() {
                None
            } else {
                Some(lines.join("\n"))
            }
        })
    }
}
